pkg/kubelet/preemption/preemption.go

```diff
@@ -0,0 +1,282 @@
+	// which are not fulfilled by all other burstable or best effort pods.
+	guarateedToEvict, err := getPodsToPreemptByDistance(guaranteedPods, requirements.subtract(append(bestEffortPods, burstablePods...)...))
+	if err != nil {
+		return nil, err
+	}
+	// Once we know the guaranteed pods being evicted, we can subtract their requirements from what we need to meet with burstable pods
+	// In similar fashion to guaranteed pods, we only need to consider requirements not met by all best effort pods.
+	burstableToEvict, err := getPodsToPreemptByDistance(burstablePods, requirements.subtract(append(bestEffortPods, guarateedToEvict...)...))
+	if err != nil {
+		return nil, err
+	}
+	bestEffortToEvict, err := getPodsToPreemptByDistance(bestEffortPods, requirements.subtract(append(burstableToEvict, guarateedToEvict...)...))
+	if err != nil {
+		return nil, err
+	}
+	return append(append(bestEffortToEvict, burstableToEvict...), guarateedToEvict...), nil
+}
+
+// finds the pods that have pod requests >= admission requirements.
+// Chooses pods that minimize "distance" to the requirements.
+// If more than one pod exists that fulfills the remaining requirements,
+// it chooses the pod that has the "smaller resource request"
+func getPodsToPreemptByDistance(pods []*v1.Pod, requirements admissionRequirementList) ([]*v1.Pod, error) {
+	podsToEvict := []*v1.Pod{}
+	// evict pods by shortest distance from remaining requirements, updating requirements every round.
+	for len(requirements) > 0 {
+		if len(pods) == 0 {
+			return nil, fmt.Errorf("no set of running pods found to reclaim resources: %v", requirements.toString())
+		}
+		// all distances must be less than len(requirements), because the max distance for a single requirement is 1
+		minDistance := float64(len(requirements) + 1)
+		minPod := 0
+		for i, pod := range pods {
+			dist := requirements.distance(pod)
+			if dist < minDistance || minDistance == 0 && dist == 0 && smallerResourceRequest(pod, pods[minPod]) {
+				minDistance = dist
+				minPod = i
+			}
+		}
+		// subtract the pod from requirements, and transfer the pod from input-pods to pods-to-evicted
+		requirements = requirements.subtract(pods[minPod])
+		podsToEvict = append(podsToEvict, pods[minPod])
+		pods[minPod] = pods[len(pods)-1]
+		pods = pods[:len(pods)-1]
+	}
+	return podsToEvict, nil
+}
+
+type admissionRequirement struct {
+	resourceName v1.ResourceName
+	quantity     int64
+}
+
+type admissionRequirementList []*admissionRequirement
+
+// distance of the pods requests from the admissionRequirements.
+// distance is measured by the fraction of the requirement satisfied by the pod,
+// so that each requirement is weighted equally, regardless of absolute magnitude.
+func (a admissionRequirementList) distance(pod *v1.Pod) float64 {
+	dist := float64(0)
+	for _, req := range a {
+		remainingRequest := float64(req.quantity - getResourceRequest(pod, req.resourceName))
+		if remainingRequest < 0 {
+			remainingRequest = 0
```

Comment:
By setting to `0` here aren't we hiding the fact that some pods might free up more resources than needed?

If we want to reclaim `1` cpu and the pod we are evaluating had requested `2` cpus, then the `dist` would be `0` IIUC. If we evaluate with a pod that had requested `1` cpu, the distance would still be `0`. 
Why not make the distance be `-1` in the previous case and `0` in the latter case? That way we can choose the evict the latter pod instead of the former.